pkg/controller/apiextensions/composite/composed.go

```diff
@@ -0,0 +1,155 @@
+func (r *APIComposedReconciler) Reconcile(ctx context.Context, cr resource.Composite, composedRef v1.ObjectReference, tmpl v1alpha1.ComposedTemplate) (Observation, error) {
+	// Deletion of the composite resource has been triggered. We make deletion
+	// the deletion call and report back success only if the call returns NotFound.
+	if meta.WasDeleted(cr) {
+		if composedRef.Name == "" {
+			return Observation{}, nil
+		}
+		err := r.client.Delete(ctx, unstructured.NewComposed(unstructured.FromReference(composedRef)))
+		if resource.IgnoreNotFound(err) != nil {
+			return Observation{}, err
+		}
+		return Observation{}, nil
+	}
+
+	var composed resource.Composable
+	if composedRef.Name == "" {
+		composed = unstructured.NewComposed()
+		if err := r.Configure(cr, composed, tmpl); err != nil {
+			return Observation{}, err
+		}
+	} else {
+		composed = unstructured.NewComposed(unstructured.FromReference(composedRef))
+		if err := r.client.Get(ctx, types.NamespacedName{Name: composed.GetName(), Namespace: composed.GetNamespace()}, composed); err != nil {
+			return Observation{}, err
+		}
+	}
+
+	// Patches are continuously applied from the Composite resource to the composed.
+	if err := r.Overlay(cr, composed, tmpl.Patches); err != nil {
+		return Observation{}, err
+	}
+
+	obs := Observation{}
+	if err := r.client.Apply(ctx, composed, resource.MustBeControllableBy(cr.GetUID())); err != nil {
+		return Observation{}, err
+	}
+	obs.Ref = *meta.ReferenceTo(composed, composed.GetObjectKind().GroupVersionKind())
+
+	conn, err := r.GetConnectionDetails(ctx, composed, tmpl.ConnectionDetails)
+	if err != nil {
+		return Observation{}, err
+	}
+	obs.ConnectionDetails = conn
+
+	return obs, nil
+}
+
+// Configure the composed object with given template and composite metadata.
+func (r *APIComposedReconciler) Configure(cr, composed resource.Object, tmpl v1alpha1.ComposedTemplate) error {
+	if err := json.Unmarshal(tmpl.Base.Raw, composed); err != nil {
+		return err
+	}
+	composed.SetGenerateName(fmt.Sprintf("%s-", cr.GetName()))
+	return nil
+}
+
+// Overlay applies an overlay to the resource with the information from parent
+// composite resource.
+func (r *APIComposedReconciler) Overlay(cr, composed resource.Object, patches []v1alpha1.Patch) error {
+	for i, patch := range patches {
+		if err := patch.Patch(cr, composed); err != nil {
+			return errors.Wrap(err, fmt.Sprintf("cannot apply the patch at index %d on result", i))
+		}
+	}
+	err := meta.AddControllerReference(composed, meta.AsOwner(meta.ReferenceTo(cr, cr.GetObjectKind().GroupVersionKind())))
+	return errors.Wrap(err, "cannot add controller ref to composed resource")
+}
+
+// GetConnectionDetails returns the ConnectionDetails of the resource if a reference
+// to the connection secret exists.
+func (r *APIComposedReconciler) GetConnectionDetails(ctx context.Context, composed resource.ConnectionSecretWriterTo, filter []v1alpha1.ConnectionDetail) (managed.ConnectionDetails, error) {
```

Comment:
It only returns the data rather than `Apply`ing it to the api-server. Later the output of this is filtered once more before publishing. Also `filter` sounds like a bad argument name since it contains information about conversion, so I changed it to `conversion`.